boxoffice/models/invoice.py

```diff
@@ -0,0 +1,93 @@
+    """
+    Returns the last invoice number used, 0 if no order has ben invoiced yet
+    """
+    query = db.session.query(sql.functions.max(Invoice.invoice_no))\
+        .filter(Invoice.organization == organization)
+    if organization.fy_start_at:
+        query = query.filter(Invoice.invoiced_at >= organization.fy_start_at)
+    last_invoice_no = query.first()
+    return last_invoice_no[0] if last_invoice_no[0] else 0
+
+
+class Invoice(UuidMixin, BaseMixin, db.Model):
+    __tablename__ = 'invoice'
+    __uuid_primary_key__ = True
+    __table_args__ = (db.UniqueConstraint('organization_id', 'invoice_no'),)
+
+    invoicee_name = db.Column(db.Unicode(255), nullable=True)
+    invoicee_email = db.Column(db.Unicode(254), nullable=True)
+    invoice_no = db.Column(db.Integer(), nullable=True)
+    invoiced_at = db.Column(db.DateTime, nullable=True)
+    street_address = db.Column(db.Unicode(255), nullable=True)
+    city = db.Column(db.Unicode(255), nullable=True)
+    state = db.Column(db.Unicode(255), nullable=True)
+    # India specific: this is the state short code. Eg: KA for Karnataka
+    state_code = db.Column(db.Unicode(4), nullable=True)
+    # ISO country code
+    country_code = db.Column(db.Unicode(2), nullable=True)
+    postcode = db.Column(db.Unicode(8), nullable=True)
+    # GSTIN in the case of India
+    buyer_taxid = db.Column(db.Unicode(255), nullable=True)
+    seller_taxid = db.Column(db.Unicode(255), nullable=True)
+
+    customer_order_id = db.Column(None, db.ForeignKey('customer_order.id'), nullable=False, index=True)
+    order = db.relationship('Order', backref=db.backref('invoices', cascade='all, delete-orphan'))
+
+    organization_id = db.Column(None, db.ForeignKey('organization.id'), nullable=False)
+    organization = db.relationship('Organization', backref=db.backref('invoices', cascade='all, delete-orphan', lazy='dynamic'))
+
+    def __init__(self, *args, **kwargs):
+        organization = kwargs.get('organization')
+        if not organization:
+            raise ValueError(u"Invoice MUST be initialized with an organization")
+        self.invoice_no = get_latest_invoice_no(organization) + 1
+        self.invoiced_at = func.utcnow()
+        super(Invoice, self).__init__(*args, **kwargs)
+
+
+@event.listens_for(Invoice, 'before_update')
+@event.listens_for(Invoice, 'before_insert')
+def validate_invoice_organization(mapper, connection, target):
+    if target.organization != target.order.organization:
+        raise ValueError(u"Invoice MUST be associated with the same organization as its order")
+
+
+class InvoiceLineItem(UuidMixin, BaseMixin, db.Model):
+    __tablename__ = 'invoice_line_item'
+    __uuid_primary_key__ = True
+    __table_args__ = (db.UniqueConstraint('invoice_id', 'seq'),)
+
+    seq = db.Column(db.Integer, nullable=False)
+    item_title = db.Column(db.Unicode(255), nullable=False)
+    quantity = db.Column(db.SmallInteger, nullable=False, default=1)
```

Comment:
`SmallInteger` is typically warranted when you have an enumeration, like status numbers.